### DISCUSSION TITLE
Run noScribe on Linux

### DISCUSSION THREAD

**Discussion by jago-jago** [kaixxx/noScribe#60] [Q&A] [answered]
What a wonderful tool! Thank you very much for this! 

I find myself checking quite often these days whether you have found the time to write some instructions on how to make noScribe work on linux. On my main computer I can make it work in a virtual machine (which puts quite a strain on my system) but aside from running it natively I would really love to love to be able to run it over night on a raspberry. 

Any idea when you might get around to looking into that? As I have no experience at all with python I am pretty lost trying to figure out how to port this to linux myself.

Again: brilliant work! Thanks!

**Comment by kaixxx** [comment]
The best instructions I have are these: https://github.com/kaixxx/noScribe/pull/33#issuecomment-1837454630
You have to install Python (version 3.9 or bigger) and Git first, if not already on your system. Then follow the instructions above, it should be quite straightforward. 
The noScribe Editor will be missing, but you can continue using it in your virtual machine. 
I'm pretty sure though that a raspberry pi is to weak for this task...

**Comment by jago-jago** [comment]
Thank you for this incredibly fast reply! It almost works!

noScribe opens up, but I'm getting the following error message:
```
Error in options
name 'get_cuda_device_count' is not defined
```
Do you have any suggestions for that?


But I got it to open up modifying your suggestion as follows (note particularly the underline in "./requirements_linux"):

```
git clone https://github.com/kaixxx/noScribe.git
cd noScribe
python3 -m venv .venv
source .venv/bin/activate
python3 -m pip install -r environments/requirements_linux.txt
python noScribe.py


```

**Reply by gernophil** [reply]
We may have implemented cuda, after making the linux modifications. Maybe this function doesn't work with the linux torch package. Or we don't have a cuda/cpu selection in the Linux version, @kaixxx? Do you have a cuda compatible GPU?

**Reply by gernophil** [reply]
this is the problem:
```
if platform.system() == 'Windows':
    ...
    from ctranslate2 import get_cuda_device_count
```
`get_cuda_device_count` only gets imported in windows. Can you try to change line 37 of `noScribe.py` from
```
if platform.system() == 'Windows':
```
to:
```
if platform.system() in ("Windows", "Linux"):
```
and try again?

**Reply by jago-jago** [reply]
Well, now it doesn't even start. I get:

```
Traceback (most recent call last):
  File "[my-home-folder]/noScribe/noScribe.py", line 39, in <module>
    from subprocess import STARTUPINFO, STARTF_USESHOWWINDOW
ImportError: cannot import name 'STARTUPINFO' from 'subprocess' (/usr/lib/python3.11/subprocess.py)

```

**Reply by gernophil** [reply]
Ah, sorry. My bad. It's late, so I missed this :). Please change this (line 37-40):
```
if platform.system() == 'Windows':
    # import torch.cuda # to check with torch.cuda.is_available()
    from subprocess import STARTUPINFO, STARTF_USESHOWWINDOW
    from ctranslate2 import get_cuda_device_count
```
to this
```
if platform.system() == 'Windows':
    # import torch.cuda # to check with torch.cuda.is_available()
    from subprocess import STARTUPINFO, STARTF_USESHOWWINDOW
if platform.system() in ("Windows", "Linux"):
    from ctranslate2 import get_cuda_device_count
```

**Reply by jago-jago** [reply]
Okay, interesting.

On my Debian 12 system I get:

```
Converting audio...
ffmpeg: [mp3 @ 0x759f780] Estimating duration from bitrate, this may be inaccurate


Canceling... (please wait a second)
ffmpeg: [mp3float @ 0x75a20c0] invalid new backstep -1

Audio conversion finished

Speaker identification...
Loading pyannote
Error in step 2 - identifiying the speakers.
cannot access local variable 'diarize_abspath' where it is not associated with a value
```
And if i try not to identify the speaker it goes on with
```

Error in step 3 - transcription
Unable to open file 'model.bin' in model '[home folder]/noScribe/models/faster-whisper-large-v2'

```
And on the raspberry I get 


```
Error in step 1 - converting the audio.
[Errno 8] Exec format error: '[my home folder]/noScribe/ffmpeg-linux-x86_64'

```


And yes: it is late! Thank you very much for all your work at this hour of night!

**Reply by gernophil** [reply, marked as answer]
> On my Debian 12 system I get:
> 
> ```
> ...
> cannot access local variable 'diarize_abspath' where it is not associated with a value
> ```

Try to change this
```
                        if platform.system() == 'Windows':
                            diarize_abspath = os.path.join(app_dir, 'diarize.exe')
                        elif platform.system() == 'Darwin': # = MAC
                            # No check for arm64 or x86_64 necessary, since the correct version will be compiled and bundled
                            diarize_abspath = os.path.join(app_dir, '..', 'MacOS', 'diarize')
                        if not ('diarize_abspath' in globals() or os.path.exists(diarize_abspath)): # Run the compiled version of diarize if it exists, otherwise the python script:
                            diarize_abspath = 'python ' + os.path.join(app_dir, 'diarize.py')
```
to this
```
                        diarize_abspath = 'python ' + os.path.join(app_dir, 'diarize.py')
```
(Remove the original lines 965 to 970 and unintend the original line 971 by one level (meaning remove 4 spaces in the beginning of the line). Remember you added a line before, so the line numbers are different for you. We did not assign `diarize_abspath` initially for linux. That's something we need to fix on our side, but this solution should be a quick workaround)

> And if i try not to identify the speaker it goes on with
> 
> ```
> 
> Error in step 3 - transcription
> Unable to open file 'model.bin' in model '[home folder]/noScribe/models/faster-whisper-large-v2'
> ```

That's because you need to download the models separately. Check the manual in `models/faster-whisper-large-v2/download.txt` and `models/faster-whisper-small/download.txt`

> And on the raspberry I get
> 
> ```
> Error in step 1 - converting the audio.
> [Errno 8] Exec format error: '[my home folder]/noScribe/ffmpeg-linux-x86_64'
> ```

The linux executable might not work with the linux distribution you are running on your raspberry. Which Linux did you install there? I guess you need a ffmpeg binary for the distro.

**Reply by jago-jago** [reply]
You are brilliant! 

That's it! Both speaker identification and transcription work nicely - and much faster compared to the virtual machine. 

 
>The linux executable might not work with the linux distribution you are running on your raspberry. >Which Linux did you install there? I guess you need a ffmpeg binary for the distro.

I have the standard raspberry os which is a Debian 12 at its core. But you might be on to something with the ffmpeg binary... But since it could not be resolved with a quick search I'll deal with it another day.

**Reply by kaixxx** [reply]
Nice to see this issue beeing resolved. (I'm on holidays...)
@gernophil: Thanks for all your help! Could you update the main codebase with the necessary changes for linux support?

**Reply by gernophil** [reply]
> Nice to see this issue beeing resolved. (I'm on holidays...)
> 
> @gernophil: Thanks for all your help! Could you update the main codebase with the necessary changes for linux support?

I'm on holidays too :D. But I'll return tomorrow and make the changes on the weekend then :).

**Reply by gernophil** [reply]
@kaixxx:
I think the line
```
if not ('diarize_abspath' in globals() or os.path.exists(diarize_abspath)):
```
does not make sense in general because we first check, if `diarize_abspath` was set or if the file exists. I think we cannot combine this with an or, because the second check will raise an error, if it's not set. This is what happened here, I guess. I think in thsi case we need a two step if clause, not an or.

**Reply by gernophil** [reply]
@jago-jago: I made some changes. Could you check, if this updated code works for you without any changes now? Also could you remove the lines 971 and 972 and check, if it would still work?

**Reply by gernophil** [reply]
Needed to make another little change since I forgot a `not`, which was very important 😅. You can test now.

**Reply by jago-jago** [reply]
Yep, it works just as the script you asked me to modify. Thank you!

Two notes:

1.  After entering the name of the html-file wherein to save the transcript it omits the point before &ldquo;html&rdquo; (eg. instead of &ldquo;transcript.html&rdquo; I get &ldquo;transcripthtml&rdquo;). If i do not correct it I get an `Invalid file type ""` error.
2.  If you intend to include an installation guide for linux, make sure to include `--use-pep517` when installing the requirements. To my mind this should get most linux users there:

 ```
   git clone https://github.com/kaixxx/noScribe.git
   cd noScribe
   # installing the models by following the link
   nano models/faster-whisper-large-v2/download.txt
   python3 -m venv .venv
   # the following line needs to be executed every time before starting noscribe
   source .venv/bin/activate
   python3 -m pip install -r environments/requirements_linux.txt --use-pep517
   # starting noscribe
   python noScribe.py

```
Lastly and coming back to the question of using a raspberry pi:

> I'm pretty sure though that a raspberry pi is to weak for this task...
-   I think it might work with the new raspberry 5 which is quite a bit more powerful

-   considering that the combination of noscribe with a raspberry pi might be attractive to other users as well, I have two questions:
    1.  Could the issue with ffmpeg be resolved through the requirements? In other words: could the ffmpeg binary be something that could be installed through python pip in the environment that is called though source? If so, how would that be done? Unfortunately, I was not able to find a solution looking into raspberry pi-related forums.
    2.  Is there a command line way of starting transcriptions?

**Reply by eckhrd** [reply]
> @kaixxx: I think the line
> 
> ```
> if not ('diarize_abspath' in globals() or os.path.exists(diarize_abspath)):
> ```
> 
> does not make sense in general because we first check, if `diarize_abspath` was set or if the file exists. I think we cannot combine this with an or, because the second check will raise an error, if it's not set. This is what happened here, I guess. I think in thsi case we need a two step if clause, not an or.

There is another issue here, namely defining a variable in the scope of a method does not make it global so

```Python
'diarize_abspath' in globals()
```

will always return `False`:

```Python
In [1]: class App(object):
   ...: 
   ...:     def define_var(self):
   ...: 
   ...:         #global foo
   ...:         foo = 'bar'
   ...:         print('foo' in globals())
   ...: 

In [2]: obj = App()

In [3]: obj.define_var()
False

In [4]: 'obj' in globals()
Out[4]: True
```

However uncommenting the `global foo` line would do the trick in this example.

I came across these issues when refactoring noScribe in my [cli branch](https://github.com/zedif/noScribe/tree/cli), where I [changed](https://github.com/zedif/noScribe/commit/12b512c0d67859f2f3e01e5aef2406ce38448750#diff-7e0c88326d87e671c14e0efd7840f14308d2dafb7215f6c7cc58f837206807b7R188-R196) the check to:

```Python
diarize_abspath = ''
if platform.system() == 'Windows':
    diarize_abspath = os.path.join(app_dir, 'diarize.exe')
if platform.system() == 'Darwin': # = MAC
    # No check for arm64 or x86_64 necessary, since the correct version will be compiled and bundled
    diarize_abspath = os.path.join(app_dir, '..', 'MacOS', 'diarize')
if not (diarize_abspath and os.path.exists(diarize_abspath)): # Run the compiled version of diarize if it exists, otherwise the python script:
    diarize_abspath = 'python ' + os.path.join(app_dir, 'diarize.py')
cfg.pyannote.abspath = diarize_abspath
```

But there might be nicer versions of doing this, too.

**Reply by gernophil** [reply]
The question is, if we need this check at all. It's only necessary, if someone runs the python script on an unsupported platform.
How bout this:
```
diarize_abspath = 'python ' + os.path.join(app_dir, 'diarize.py')
diarize_abspath_win = os.path.join(app_dir, 'diarize.exe')
diarize_abspath_mac = os.path.join(app_dir, '..', 'MacOS', 'diarize')
diarize_abspath_lin = os.path.join(app_dir, 'diarize')
if platform.system() == 'Windows' and os.path.exists(diarize_abspath_win):
    diarize_abspath = diarize_abspath_win
elif platform.system() == 'Darwin' and os.path.exists(diarize_abspath_mac): # = MAC
    diarize_abspath = diarize_abspath_mac
elif platform.system() == 'Linux' and os.path.exists(diarize_abspath_lin):
    diarize_abspath = diarize_abspath_lin
```
EDIT: I think `elif` makes more sense here.

**Reply by eckhrd** [reply]
Yes, looks good and is also more readable in my opinion.

**Reply by jago-jago** [reply]
btw, I figured out how to make it work - just in case anyone is looking to make things run on a raspberry pi.

> considering that the combination of noscribe with a raspberry pi might be attractive to other users as well, I have two questions:
> 
>     Could the issue with ffmpeg be resolved through the requirements? In other words: could the ffmpeg binary be something that could be installed through python pip in the environment that is called though source? If so, how would that be done? Unfortunately, I was not able to find a solution looking into raspberry pi-related forums.

There is a ffmpeg binary for the raspberry pi (`sudo apt-get update; sudo apt-get install ffmpeg`)

all I needed to do is change the ffmpeg_path in noScribe.py to where it actually is (`whereis ffmpeg`) so instead of 
`                        ffmpeg_path = os.path.join(app_dir, 'ffmpeg-linux-x86_64')`
I would put 
`                        ffmpeg_path = '/usr/bin/ffmpeg'`
>     Is there a command line way of starting transcriptions?
I am very sorry for not doing my research on this! You are already working on it. I got too excited with your incredibly fast and useful feedback.

**Reply by kaixxx** [reply]
So noScribe is now working on a raspi? That's crazy. How long does it take to transcribe an audio?

**Reply by jago-jago** [reply]
Well... got a bit too excited again...

It ended up crashing while trying to load wisper. Here's what I get from the terminal:

```
~/noScribe/.venv/lib/python3.11/site-packages/pyannote/audio/core/io.py:43: UserWarning: torchaudio._backend.set_audio_backend has been deprecated. With dispatcher enabled, this function is no-op. You can remove the function call.

  torchaudio.set_audio_backend("soundfile")

torchvision is not available - cannot save figures

Lightning automatically upgraded your loaded checkpoint from v1.5.4 to v2.2.5. To apply the upgrade to your files permanently, run `python -m pytorch_lightning.utilities.upgrade_checkpoint models/pytorch_model.bin`

Model was trained with pyannote.audio 0.0.1, yours is 2.1.1. Bad things might happen unless you revert pyannote.audio to 0.x.

Model was trained with torch 1.10.0+cu102, yours is 2.3.0. Bad things might happen unless you revert torch to 1.x.

progress segmentation 0
```

[...]

```
progress segmentation 100

progress speaker_counting 100

progress embeddings 4
```

[...]

```
progress embeddings 100

progress discrete_diarization 100

killed
```

Do you think it might just be the lack of power or could there be something I'm missing?

**Reply by gernophil** [reply]
> The question is, if we need this check at all. It's only necessary, if someone runs the python script on an unsupported platform. How bout this:
> 
> ```
> diarize_abspath = 'python ' + os.path.join(app_dir, 'diarize.py')
> diarize_abspath_win = os.path.join(app_dir, 'diarize.exe')
> diarize_abspath_mac = os.path.join(app_dir, '..', 'MacOS', 'diarize')
> diarize_abspath_lin = os.path.join(app_dir, 'diarize')
> if platform.system() == 'Windows' and os.path.exists(diarize_abspath_win):
>     diarize_abspath = diarize_abspath_win
> elif platform.system() == 'Darwin' and os.path.exists(diarize_abspath_mac): # = MAC
>     diarize_abspath = diarize_abspath_mac
> elif platform.system() == 'Linux' and os.path.exists(diarize_abspath_lin):
>     diarize_abspath = diarize_abspath_lin
> ```
> 
> EDIT: I think `elif` makes more sense here.

I pushed this and checked it. For Mac it still works. For the others it should too.

**Reply by kaixxx** [reply]
@jago-jago: Hard to tell what the problem is since there is no real error message. Try to run it with speaker detection set to "none". This way we can see if whisper even works.
